File: fixtures/code/shadowing.rb
# Every case the Ruby vocabulary has to get right, in one file.
#
# Read by src/ledger/dialects/ruby.test.ts and by a person. Two things here are
# Ruby's own: a method is called with no receiver, and a `for` leaves its name
# behind where a block does not.
require 'set'
require_relative 'local/helper'

LIMIT = 10

if LIMIT > 1
  leaked = 1
end

module Shapes
  class Box < Base
    attr_reader :size
    SIDES = 4

    def initialize(size)
      @size = size
    end

    def draw(n)
      total = n * @size
      [1, 2].each { |row| total += row }
      for i in 0..3
        total += i
      end
      begin
        total += risky
      rescue StandardError => err
        return 0
      end
      total
    end

    def self.build(n) = new(n)

    private

    def risky
      LIMIT
    end
  end

  Kind = Struct.new(:name)
end

doubled = ->(z) { z * 2 }
first, second = 1, 2

File: fixtures/code/shadowing.php
<?php
// Every case the PHP vocabulary has to get right, in one file.
//
// Read by src/ledger/dialects/php.test.ts and by a person. Two things here are
// PHP's own: a variable is a name inside a name, and binding happens at the
// function rather than at the block.
namespace App\Shapes;

use App\Base;
use App\Other\Thing as Widget;
use App\Grouped\{Alpha, Beta};

const LIMIT = 10;

function area(int $shape, int $scale = 1): int {
    $total = $shape * $scale;
    foreach ([1, 2] as $row) { $total += $row; }
    for ($i = 0; $i < 3; $i++) { $total += $i; }
    try { $total += risky(); }
    catch (Exception $err) { return 0; }
    $doubled = fn($z) => $z * 2;
    return $total + $doubled(1);
}

function risky(): int { return LIMIT; }

class Box extends Base {
    private int $size = 4;
    const SIDES = 4;

    public function __construct(int $size) { $this->size = $size; }

    public function draw(int $n): int { return area($n, $this->size); }
}

interface Shape { public function area(): int; }

trait Drawable { public function drawn(): bool { return true; } }
